### PR TITLE
[Exporter Refactor] `BaseTransform` and `OnnxTransform`

### PR DESCRIPTION
**Pull request by dbogunowicz**
This PR adds two interfaces: `BaseTransform` and `OnnxTransform` (that inherits from `BaseTransform`)

### CODE REVIEW

File: src/sparseml/exporters/transforms/base_transform.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    @abstractmethod
+    def transform(self, model: Any) -> Any:
+        """
+        Logic for applying the transformation to the model.
+
+        :param model: The input model to be transformed
+        :return: The transformed model
+        """
+        # The transform algorithm that will be applied to the model
```
Comment (dbogunowicz) [inline src/sparseml/exporters/transforms/base_transform.py:53]:
```suggestion
        # The transform algorithm that will be applied to the model
```

File: src/sparseml/exporters/transforms/base_transform.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        should follow the following steps:
+        1. Validate the input model
+        2. Apply the transform to the model
+        3. Validate the resulting model and return it
+
+        :param model: The input model to be
+            validated, transformed and once again, validated
+        :return: The transformed model
+        """
+        raise NotImplementedError
```
Comment (corey-nm) [inline src/sparseml/exporters/transforms/base_transform.py:43]: optional nit: default implementation
```suggestion
        self.pre_validate(model)
        self.transform(model)
        self.post_validate(model)
```

File: src/sparseml/exporters/transforms/base_transform.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    Interface that all transforms must implement.
+    A transform is an object that, when applied to a model,
+    does a single modification to it
+    and returns a modified model.
+    """
+
+    def __call__(self, model: Any) -> Any:
+        return self.apply(model)
+
+    @abstractmethod
```
Comment (corey-nm) [inline src/sparseml/exporters/transforms/base_transform.py:30]: if default implementation comment below, then remove this

File: src/sparseml/exporters/transforms/onnx_transform.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            It can be a path to the model or the model itself.
+        :return: The transformed ONNX model
+        """
+        if not ((isinstance(model, str) or isinstance(model, ModelProto))):
+            raise ValueError(
+                f"Invalid model type: {type(model)}. "
+                "Must be a string (path to the .onnx file) or ONNX ModelProto"
+            )
+
+        model = check_load_model(model)
```
Comment (corey-nm) [inline src/sparseml/exporters/transforms/onnx_transform.py:50]: move these to pre_validate maybe?

Also maybe make pre_validate/post_validate return models as well?

If both of these then i think we don't need to override the default implementation of apply above

**Review by corey-nm — APPROVED** [review]
looks good to me! left some small suggestions, but this is great 🔥

File: src/sparseml/exporters/transforms/base_transform.py (hunk 2 above)
Comment (bfineran) [reply src/sparseml/exporters/transforms/base_transform.py:43]: +1

Comment (bfineran) [reply src/sparseml/exporters/transforms/base_transform.py:43]: additionally, what's the expected flow here? do the validation functions raise errors as they encounter them, or do we return an error string that the apply step will raise?

**Review by bfineran — APPROVED** [review]
LGTM pending comment

**Pull request merged**